posts/2019-07-04-Rust-1.36.0.md

```diff
@@ -0,0 +1,164 @@
+
+As [`MaybeUninit<T>`] is the safer alternative, starting with Rust 1.38, the function [`mem::uninitialized`] will be deprecated.
+
+To read more about uninitialized memory, [`mem::uninitialized`], and [`MaybeUninit<T>`], read [Alexis Beingessner's blog post][gankro-blog] about the subject. The standard library also contains extensive documentation about [`MaybeUninit<T>`].
+
+### NLL for Rust 2015
+
+[nll-2018]: https://blog.rust-lang.org/2018/12/06/Rust-1.31-and-rust-2018.html#non-lexical-lifetimes
+[soundness]: https://en.wikipedia.org/wiki/Soundness
+[felix-blog]: http://blog.pnkfx.org/blog/2019/06/26/breaking-news-non-lexical-lifetimes-arrives-for-everyone/
+[crater-nll]: https://github.com/rust-lang/rust/issues/60680#issuecomment-495089654
+
+[In the announcement for Rust 1.31.0][nll-2018], we told you about NLL (Non-Lexical Lifetimes), an improvement to the language that makes the borrow checker smarter and more user friendly. For example, you may now write:
+
+```rust
+fn main() {
+    let mut x = 5;
+    let y = &x;
+    let z = &mut x; // This was not allowed before 1.31.0.
+}
+```
+
+In 1.31.0 NLL was stabilized only for Rust 2018, with a promise that we would backport it to Rust 2015 as well. With Rust 1.36.0, we are happy to announce that we have done so! NLL is now available for Rust 2015.
+
+With NLL on both editions, we are closer to removing the old borrow checker. However, the old borrow checker unfortunately accepted some [unsound][soundness] code it should not have. As a result, NLL is currently in a "migration mode" wherein we will emit warnings instead of errors if the NLL borrow checker rejects code the old AST borrow checker would accept. Please see [this list][crater-nll] of public crates that are affected.
+
+To read more about NLL, MIR, the story around fixing soundness holes, and what you can do about the warnings if you have them, [read Felix Klock's blog post][felix-blog].
+
+### A new `HashMap<K, V>` implementation
+
+[`hashbrown`]: https://crates.io/crates/hashbrown
+[`HashMap<K, V>`]: https://doc.rust-lang.org/std/collections/struct.HashMap.html
+[pr-hashbrown]: https://github.com/rust-lang/rust/pull/58623
+[SwissTable]: https://abseil.io/blog/20180927-swisstables
+[pr-hashbrown-perf]: https://perf.rust-lang.org/compare.html?start=b57fe74a27590289fd657614b8ad1f3eac8a7ad2&end=abade53a649583e40ed07c26ee10652703f09b58&stat=wall-time
+
+In Rust 1.36.0, the `HashMap<K, V>` implementation has been [replaced][pr-hashbrown] with the one in the [`hashbrown`] crate which is based on the [SwissTable] design. While the interface is the same, the `HashMap<K, V>` implementation is now [faster on average][pr-hashbrown-perf] and has lower memory overhead. Note that unlike the `hashbrown` crate, the implementation in `std` continues to use the SipHash 1-3 hashing algorithm.
```

Comment:
Maybe “continues to use the SipHash 1-3 hashing algorithm by default”, or “continues to default to the SipHash 1-3 hashing algorithm”? `std::collections::HashMap` supports custom hashers.

Reply:
Maybe ", the implementation in `std` still defaults to the SipHash 1-3 hashing algorithm." ?

Reply:
Yes, that works.

Reply:
Cool;
```suggestion
In Rust 1.36.0, the `HashMap<K, V>` implementation has been [replaced][pr-hashbrown] with the one in the [`hashbrown`] crate which is based on the [SwissTable] design. While the interface is the same, the `HashMap<K, V>` implementation is now [faster on average][pr-hashbrown-perf] and has lower memory overhead. Note that unlike the `hashbrown` crate, the implementation in `std` still defaults to the SipHash 1-3 hashing algorithm.
```